stan/math/prim/prob/hmm_marginal_lpdf.hpp

```diff
@@ -0,0 +1,234 @@
+                        Eigen::Matrix<T_Gamma, Eigen::Dynamic, Eigen::Dynamic>,
+                        Eigen::Matrix<T_rho, Eigen::Dynamic, 1> >
+      ops_partials(log_omegas, Gamma, rho);
+
+  Eigen::MatrixXd alphas(n_states, n_transitions + 1);
+  Eigen::VectorXd alpha_log_norms(n_transitions + 1);
+  Eigen::MatrixXd omegas;
+  Eigen::MatrixXd Gamma_dbl = value_of_rec(Gamma);
+
+  T_partials_return log_marginal_density
+      = hmm_marginal_lpdf(value_of_rec(log_omegas), Gamma_dbl,
+                          value_of_rec(rho), alphas, alpha_log_norms, omegas);
+
+  // Variables required for all three Jacobian-adjoint products.
+  double norm_norm = alpha_log_norms(n_transitions);
+  double unnormed_marginal = alphas.col(n_transitions).sum();
+
+  std::vector<Eigen::VectorXd> kappa(n_transitions);
+  Eigen::VectorXd kappa_log_norms(n_transitions);
+  std::vector<double> grad_corr(n_transitions);
+
+  if (n_transitions > 0) {
+    kappa[n_transitions - 1] = Eigen::VectorXd::Ones(n_states);
+    kappa_log_norms(n_transitions - 1) = 0;
+    grad_corr[n_transitions - 1]
+        = std::exp(alpha_log_norms(n_transitions - 1) - norm_norm);
+  }
+
+  for (int n = n_transitions - 2; n >= 0; --n) {
+    kappa[n] = Gamma_dbl * (omegas.col(n + 2).cwiseProduct(kappa[n + 1]));
+
+    double norm = kappa[n].maxCoeff();
+    kappa[n] /= norm;
+    kappa_log_norms(n) = std::log(norm) + kappa_log_norms(n + 1);
+    grad_corr[n]
+        = std::exp(alpha_log_norms(n) + kappa_log_norms(n) - norm_norm);
+  }
+
+  if (!is_constant_all<T_Gamma>::value) {
+    Eigen::MatrixXd Gamma_jacad(n_states, n_states);
+    Gamma_jacad.setZero();
+
+    if (n_transitions != 0) {
+      for (int n = n_transitions - 1; n >= 0; --n) {
+        Gamma_jacad += (grad_corr[n] * kappa[n].cwiseProduct(omegas.col(n + 1))
+                        * alphas.col(n).transpose())
+                           .transpose();
+      }
+    }
+
+    Gamma_jacad /= unnormed_marginal;
+    ops_partials.edge2_.partials_ = Gamma_jacad;
+  }
+
+  bool sensitivities_for_omega_or_rho
+      = (!is_constant_all<T_omega>::value) || (!is_constant_all<T_rho>::value);
+
+  // boundary terms
+  if (sensitivities_for_omega_or_rho) {
+    Eigen::MatrixXd log_omega_jacad(n_states, n_transitions + 1);
+    log_omega_jacad.setZero();
+
+    if (!is_constant_all<T_omega>::value) {
+      for (int n = n_transitions - 1; n >= 0; --n)
+        log_omega_jacad.col(n + 1)
+            = grad_corr[n]
+              * kappa[n].cwiseProduct(Gamma_dbl.transpose() * alphas.col(n));
+    }
+
+    // Boundary terms
+    // TO DO (charlesm93): find a better solution that the if loop
+    // for the case with 0 transitions.
+    double grad_corr_boundary;
+    Eigen::VectorXd c;
+    if (n_transitions != 0) {
+      grad_corr_boundary = std::exp(kappa_log_norms(0) - norm_norm);
+      c = Gamma_dbl * omegas.col(1).cwiseProduct(kappa[0]);
+    }
+
+    if (!is_constant_all<T_omega>::value) {
+      if (n_transitions != 0) {
```

Comment:
[optional]

I'd prefer something at the very top handling the `n_transitions == 0` case so it's not in the way of understanding all the code.  (Overall the code's really understandable.)

Reply:
I agree. I'll work on it.

Reply:
+1, also Charles note that the if statements that use type traits like `is_constant_all` are resolved at compile time (i.e. it just becomes `if (true/false)`) so the alt code path will just be deleted since it's dead so you don't need to worry about those ifs in terms of performance

Reply:
I think I found a fairly good solution. The `n_transitions == 0` case only matters when handling the boundary terms, so it's now handled at the very top, when we start treating the boundary terms.